tests/extra-tests/src/test/java/org/apache/activemq/artemis/tests/extras/byteman/RaceOnCursorIteratorTest.java

```diff
@@ -0,0 +1,208 @@
+
+   static boolean skipLivePageCache = false;
+
+   static boolean skipNullPageCache = false;
+
+   static boolean moveNextPageCalled = false;
+
+   @Override
+   @Before
+   public void setUp() throws Exception {
+      super.setUp();
+      skipLivePageCache = false;
+      skipNullPageCache = false;
+      moveNextPageCalled = false;
+      locator = createInVMNonHALocator();
+
+      clearDataRecreateServerDirs();
+
+      Configuration config = createDefaultConfig(false);
+
+      config.setJournalSyncNonTransactional(false);
+
+      HashMap<String, AddressSettings> map = new HashMap<>();
+      AddressSettings value = new AddressSettings();
+      map.put(ADDRESS.toString(), value);
+      server = createServer(true, config, PAGE_SIZE, PAGE_MAX, map);
+
+      server.start();
+
+      locator = createInVMNonHALocator();
+
+      locator.setBlockOnNonDurableSend(true);
+      locator.setBlockOnDurableSend(true);
+      locator.setBlockOnAcknowledge(false);
+      locator.setConsumerWindowSize(0);
+
+      sf = createSessionFactory(locator);
+
+      session = sf.createSession(false, true, true);
+
+      session.createQueue(ADDRESS, ADDRESS, null, true);
+
+      queue = server.locateQueue(ADDRESS);
+      queue.getPageSubscription().getPagingStore().startPaging();
+   }
+
+   @Override
+   @After
+   public void tearDown() throws Exception {
+      session.close();
+      sf.close();
+      locator.close();
+      server.stop();
+      super.tearDown();
+   }
+
+   public static void raceAddLivePageCache() throws Exception {
+      if (skipLivePageCache) {
+         createMessage(1);
+
+         queue.getPageSubscription().getPagingStore().forceAnotherPage();
+
+         createMessage(2);
+      }
+      moveNextPageCalled = true;
+   }
+
+   public static void raceAddTwoCaches() throws Exception {
+      if (skipNullPageCache && moveNextPageCalled) {
+         createMessage(1);
+
+         queue.getPageSubscription().getPagingStore().forceAnotherPage();
+
+         createMessage(2);
+      }
+   }
+
+   @Test
+   @BMRules(
+      rules = {@BMRule(
+         name = "raceLiveCache",
+         targetClass = "org.apache.activemq.artemis.core.paging.cursor.impl.PageSubscriptionImpl",
+         targetMethod = "moveNextPage",
+         targetLocation = "EXIT",
+         action = "org.apache.activemq.artemis.tests.extras.byteman.RaceOnCursorIteratorTest.raceAddLivePageCache()")})
+   public void testSkipLivePageCache() {
+      skipLivePageCache = true;
+      // Simulate scenario #1 depicted in https://issues.apache.org/jira/browse/ARTEMIS-2418
+      PagedReference ref = queue.getPageSubscription().iterator().next();
+      assertTrue("first msg should not be " + (ref == null ? "null" : ref.getPagedMessage().getMessage().getMessageID()),
+                 ref == null || ref.getPagedMessage().getMessage().getMessageID() == 1);
+   }
+
+   @Test
+   @BMRules(
+      rules = {@BMRule(
```

Comment:
Did you really need the byteman test? These test tend to be fragile. I have been personally avoiding them.


Can you explain in english what the test is doing? I coudl read the byteman rule. but I got confused on what is the race.

Reply:
Test testSkipLivePageCache simulates scenario 1. 
> Meanwhile p1 is filled with message m1 and p2 is created with m2.
This is achieved by calling raceAddLivePageCache after entry of PageSubscriptionImpl::moveNextPage.

Test testSkipNullPageCache simulates scenario 2. 
> When we call internalGetNext(), the cursor moves to position(p2, 0) and get null page cache since p2 is not yet created. Then p2 is created with m1 and p3 is created with m2 which means current writing page no. is p3
This is achieved by calling raceAddTwoCaches after the first PageSubscriptionImpl::moveNextPage and after PageSubscriptionImpl::getPageCache.

I have added the code line in the previous comment. Hope this helped to understand the race.

I wrote the byteman test to justify the race conditions. As you said, the test is tricky and fragile, so it's ok for me to remove it :)

Reply:
I'm ok on merging this, however this will clash with your other changes on paging.


Would you rebase after I merged this?

Reply:
Yes, I would rebase, feel free to merge it :)